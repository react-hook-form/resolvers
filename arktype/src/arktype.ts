import { FieldError, FieldErrors } from 'react-hook-form';
import { toNestError, validateFieldsNatively } from '@hookform/resolvers';
import type { Resolver } from './types';
import { Problems } from 'arktype';

const parseErrorSchema = (e: Problems) => {
  const errors: Record<string, FieldError> = {};
  for (; e.length; ) {
    const error = e[0];
    const _path = error.path.join('.');

    if (!errors[_path]) {
      errors[_path] = { message: error.message, type: error.code };
    }

    // @ts-expect-error - false positive Property 'shift' does not exist on type 'Problems'.
    e.shift();
  }

  return errors;
};

export const arktypeResolver: Resolver =
  (schema, _schemaOptions, resolverOptions = {}) =>
  (values, _, options) => {
    const result = schema(values);

    if (result.problems) {
      return {
        values: {},
        errors: toNestError(parseErrorSchema(result.problems), options),
      };
    }

    options.shouldUseNativeValidation && validateFieldsNatively({}, options);

    return {
      errors: {} as FieldErrors,
      values: resolverOptions.raw ? values : result.data,
    };
  };
