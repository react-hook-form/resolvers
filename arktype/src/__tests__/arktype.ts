import { type } from 'arktype';
import { Resolver, useForm } from 'react-hook-form';
import { SubmitHandler } from 'react-hook-form';
import { arktypeResolver } from '..';
import { fields, invalidData, schema, validData } from './__fixtures__/data';

const shouldUseNativeValidation = false;

describe('arktypeResolver', () => {
  it('should return values from arktypeResolver when validation pass & raw=true', async () => {
    const result = await arktypeResolver(schema, undefined, {
      raw: true,
    })(validData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toEqual({ errors: {}, values: validData });
  });

  it('should return a single error from arktypeResolver when validation fails', async () => {
    const result = await arktypeResolver(schema)(invalidData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toMatchSnapshot();
  });

  /**
   * Type inference tests
   */
  it('should correctly infer the output type from a arktype schema', () => {
    const resolver = arktypeResolver(type({ id: 'number' }));

    expectTypeOf(resolver).toEqualTypeOf<
      Resolver<{ id: number }, unknown, { id: number }>
    >();
  });

  it('should correctly infer the output type from a arktype schema using a transform', () => {
    const resolver = arktypeResolver(
      type({ id: type('string').pipe((s) => Number.parseInt(s)) }),
    );

    expectTypeOf(resolver).toEqualTypeOf<
      Resolver<{ id: string }, unknown, { id: number }>
    >();
  });

  it('should correctly infer the output type from a arktype schema for the handleSubmit function in useForm', () => {
    const schema = type({ id: 'number' });

    const form = useForm({
      resolver: arktypeResolver(schema),
    });

    expectTypeOf(form.watch('id')).toEqualTypeOf<number>();

    expectTypeOf(form.handleSubmit).parameter(0).toEqualTypeOf<
      SubmitHandler<{
        id: number;
      }>
    >();
  });

  it('should correctly infer the output type from a arktype schema with a transform for the handleSubmit function in useForm', () => {
    const schema = type({ id: type('string').pipe((s) => Number.parseInt(s)) });

    const form = useForm({
      resolver: arktypeResolver(schema),
    });

    expectTypeOf(form.watch('id')).toEqualTypeOf<string>();

    expectTypeOf(form.handleSubmit).parameter(0).toEqualTypeOf<
      SubmitHandler<{
        id: number;
      }>
    >();
  });
});
