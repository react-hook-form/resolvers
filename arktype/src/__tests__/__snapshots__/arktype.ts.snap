// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`arktypeResolver > should return a single error from arktypeResolver when validation fails 1`] = `
{
  "errors": {
    "accessToken": {
      "message": "accessToken must be defined",
      "ref": undefined,
      "type": "missing",
    },
    "birthYear": {
      "message": "birthYear must be a number (was string)",
      "ref": undefined,
      "type": "domain",
    },
    "dateStr": {
      "message": "dateStr must be defined",
      "ref": undefined,
      "type": "missing",
    },
    "email": {
      "message": "email must be a valid email (was '')",
      "ref": {
        "name": "email",
      },
      "type": "regex",
    },
    "enabled": {
      "message": "enabled must be defined",
      "ref": undefined,
      "type": "missing",
    },
    "like": [
      {
        "id": {
          "message": "like/0/id must be a number (was string)",
          "ref": undefined,
          "type": "domain",
        },
        "name": {
          "message": "like/0/name must be defined",
          "ref": undefined,
          "type": "missing",
        },
      },
    ],
    "password": {
      "message": "At password, '___' must be...
• more than 8 characters
• a string matching /.*[A-Za-z].*/",
      "ref": {
        "name": "password",
      },
      "type": "multi",
    },
    "repeatPassword": {
      "message": "repeatPassword must be defined",
      "ref": undefined,
      "type": "missing",
    },
    "tags": {
      "message": "tags must be defined",
      "ref": undefined,
      "type": "missing",
    },
    "username": {
      "message": "username must be defined",
      "ref": {
        "name": "username",
      },
      "type": "missing",
    },
  },
  "values": {},
}
`;
